LLBLGen 5.4 and SybaseASA fixes

bulldetektor opened the pull request:
Updated to compile against LLBLGen 5.4 and made som fixes regarding templates and driver ids (mixup between Sybase ASA and ASE drivers)

Note; Please take a look at the code changes I had to do to make it compile agains 5.4. I used the SybaseASE repo for some guidelines, but not sure if I made the right decisions...

FransBouma commented on Driver/SybaseAsaDBDriver.cs:
```diff
@@ -53,7 +53,7 @@ public class SybaseAsaDBDriver : DBDriverBase
 		private const string driverVendor = "Solutions Design bv";
 		private const string driverCopyright = "(c)2002-2016 Solutions Design, all rights reserved.";
-		private const string driverID = "3FABDE1A-21DF-4fcb-96FD-BBFA8F18B1EA";
+		private const string driverID = "A3076322-977C-4e28-BFF4-F25ED096D1DB";
```
This isn't correct, you can't change the ID as that would tie the driver to a different database. Here you specify the ASE guid with this driver, so that will cause problems. Please leave the original Guid as-is.

FransBouma commented on Driver/driver.config:
```diff
@@ -8,7 +8,7 @@
 	<namespace>SD.LLBLGen.Pro.DBDrivers.SybaseAsa</namespace>
 	<className>SybaseAsaDBDriver</className>
-	<driverID>3FABDE1A-21DF-4fcb-96FD-BBFA8F18B1EA</driverID>
+	<driverID>A3076322-977C-4e28-BFF4-F25ED096D1DB</driverID>
```
Same here, just leave the original Guid value. I think you did this to make the wrong templatebindings file we had in place (which was for ASE) to work? The original guid should work with the new templatebindings file as that templatebindingsfile contains the 3FABD... guid value

FransBouma commented on DynamicQueryEngine/DynamicQueryEngine.cs:
```diff
@@ -111,24 +111,27 @@ static DynamicQueryEngine()
 			QueryFragments fragments = new QueryFragments();
 			fragments.AddFormatted("INSERT INTO {0}", this.Creator.CreateObjectName(fieldsPersistenceInfo[0]));
-			DelimitedStringList fieldNames = fragments.AddCommaFragmentList(true);
+			//DelimitedStringList fieldNames = fragments.AddCommaFragmentList(true);
```
Please remove commented out code (which is redundant now, it's present below this comment)

bulldetektor replied:
Off course! Sloppy commit :)

FransBouma commented on DynamicQueryEngine/DynamicQueryEngine.cs:
```diff
@@ -140,13 +143,13 @@ static DynamicQueryEngine()
 					// a table with just 1 identity field, use a special case query: INSERT INTO table values ()
-					fieldNames.Clear();
-					valueFragments.Clear();
+					//fieldNames.Clear();
```
Not sure why you commented this out?

bulldetektor replied:
This is the reason;

'QueryFragments' does not contain a definition for 'Clear' and no extension method 'Clear' accepting a first argument of type 'QueryFragments' could be found 

Not sure what the alternative should be? I just looked at the ASE repo and found that no 'Clear'-method was called there

FransBouma replied:
These clear calls are not needed. It builds a stack of objects in which strings are stored which is then flattened into a single buffer, to avoid unnecessary concats and inserts. Then this stack is discarded, so clear isn't needed.

FransBouma commented on Templates/SD.TemplateBindings.SybaseAsaSpecific.templatebindings:
```diff
@@ -1,6 +1,6 @@
 ﻿<?xml version="1.0"?>
 <templateBindings xmlns="http://sd/llblgen/pro/templateBindingsDefinition.xsd" name="SD.TemplateBindings.SybaseAsaSpecific"
-				  description ="Template bindings for the templates which are specific for Sybase ASA iAnywhere databases."
+				  description ="Template bindings for the templates which are specific for Sybase ASE databases."
```
I think this isn't correct, it's for asa, not ase (Admitted, it confuses me the every time as well ;))

bulldetektor replied:
yes :)

bulldetektor commented:
Regarding the driver ID; I got the guid from the ReadMe, so if that was incorrect then the ReadMe should be updated as well.

> The ID for Sybase ASA is: A3076322-977C-4e28-BFF4-F25ED096D1DB

FransBouma commented:
Absolutely correct. There has been some mixup in the filenames and files with these two repos. From the v4.2 codebase, the ID for ASA is the 3F one, for ASE the A30... one. 

There's one thing I overlooked, you removed https://github.com/SolutionsDesign/LLBLGenProSybaseASA/pull/3/files#diff-fb0e9b85768d524e98a74ce03a47d02eL243 CreateParameterName, however the default one prefixes the name with an empty string (as the property is empty by default). I think you have to leave this method in the code to get properly prefixes parameter names? 

For the rest it's OK. :) I'll clean up the ASE repo as well so things are in the right folder.

bulldetektor commented:
The reason I removed the `CreateParameterName` is that it seems to have been deprecated? I get this error if I leave it and compile it against 5.4;

> 'SybaseAsaSpecificCreator.CreateParameterName()': cannot override inherited member 'DbSpecificCreatorBase.CreateParameterName()' because it is not marked virtual, abstract, or override

bulldetektor commented:
I'm getting closer to get it up and running, but I'm facing an error that I can't figure out (or haven't had any googling luck with so far). When I'm running linq queries I can't get LLBLGen to add brackets around the column names. E.g. this query;

```
var publishedReports = db.W300Report
					.Where(
						report =>
							report.IsDeleted == false
							&& report.IsPublished)
					.ToList();
```

Should have generated something like this;

`SELECT DISTINCT (LPA_L2.[Comment], LPA_L2.[HtmlTemplate], LPA_L2 ....`

But from the profiler I can see that the generated query looks like this:

`SELECT DISTINCT (LPA_L2.Comment, LPA_L2.HtmlTemplate, LPA_L2`

Note the missing `[]` around the column names. And since 'Comment' is a protected keyword in SQL Anywhere I get a database error on the select statement.

FransBouma commented:
> The reason I removed the CreateParameterName is that it seems to have been deprecated? I get this error if I leave it and compile it against 5.4;

Crap, my brain wasn't totally booted this morning. Sorry for this... You need to add this override to SybaseAsaSpecificCreator

```cs
        /// <summary>
        /// Gets the parameter prefix, if required. If no parameter prefix is required, this property will return the empty string (by default it returns the empty string).
        /// </summary>
        protected override string ParameterPrefix
        {
            get { return "@"; }
        }
```

The problems with not having `[]` around names is due to this method: https://github.com/CapitechAS/LLBLGenProSybaseASA/blob/373bd15feec39d0732858f92a80c1ea24d13957b/DynamicQueryEngine/SybaseAsaSpecificCreator.cs#L229

It should be something like this (this is from the SQL Server DQE, so this should work for you too)
```cs
/// <inheritdoc/>
public override void AppendValidIdentifier(QueryFragments toAppendTo, string rawIdentifier)
{
	if(string.IsNullOrEmpty(rawIdentifier))
	{
		return;
	}
	if(rawIdentifier[0] == '[')
	{
		toAppendTo.AddFragment(rawIdentifier);
	}
	else
	{
		toAppendTo.AddStringFragmentsAsSingleUnit("[", rawIdentifier, "]");
	}
}
```

This is different from earlier versions as it now uses pre-allocated memory in buffers rotated in a pool instead of concatenating things all over the place.

bulldetektor commented:
Thanks! Now I'm getting the square brackets for column names, but SQLAnywhere complains about those brackets around the columns in the select statement. E.g.

`SELECT DISTINCT ([LPA_L2].[Comment], [LPA_L2].[HtmlTemplate], ...) `

Should have been:

`SELECT DISTINCT [LPA_L2].[Comment], [LPA_L2].[HtmlTemplate], ...`

bulldetektor commented:
Sorry - my bad! In an attempt to fix the square brackets for the column names I changed the CreateSelectDQ from;

`var projection = fragments.AddCommaDelimitedQueryFragments(false, 0);`

to:

`var projection = fragments.AddCommaDelimitedQueryFragments(true, 0);`

Reverted the change and it seems promising so far :)

FransBouma commented:
:) Yes that change indeed makes it to wrap the list in brackets. If you download the sourcecode archive from the website (My account -> Downloads -> v5.4->Extras section) you get the sourcecode of the runtime which contains the code for classes like this one. Also it's beneficial to pass the selectlist length to that method instead of 0, as with 0 it always has to resize which is an allocation. This is minor of course, but every allocation that's unnecessary is one too many ;) 

```cs
var projection = fragments.AddCommaDelimitedQueryFragments(false, selectList?.Length ?? 0);
```

FransBouma commented:
It looks good so far. If things all pass on your end, I'll merge it.

bulldetektor commented:
Great! Will do some more extensive testing tomorrow and let you know :)

bulldetektor commented:
Did some testing on my local machine today and it looks good so far. Need to do some more testing in test and staging environment before I give all thumbs up :)

bulldetektor commented:
Got some testing done today in a couple of different environments and it all looks good :)

FransBouma commented:
Great! Could you remove the binary files though? ORMSupportClasses are on NuGet, and the designer dlls are not to be added here (they come with the designer so should be referenced from there). Then I'll merge it.

bulldetektor commented:
Sure, I'll use the nuget version for ORMSupportClasses, but where do I get SD.LLBLGen.Pro.DBDriverCore and SD.LLBLGen.Pro.Core from? The current references were relative paths into the installed LLBLGen Pro folder (e.g. "c:\program files...."). But those references will not be correct for someone wanting to fork/clone this repo. Maybe you could publish those to nuget as well?

FransBouma commented:
The dlls referenced will not be in the repository for someone who clones it but that's ok, they have to reference the assemblies from the local designer install, so that'll make it build. (it's in the readme, so it's fine :))

bulldetektor commented:
Updated the refs, but made the SD.LLBLGen.Pro assemblies point to an absolute path (instead of the prior relative path) to the LLBLGen 5.4 installation folder. Probably more likely to cause least amount of friction for the next who'll want to do changes to the source code.

FransBouma commented:
Excellent work Kjetil :)
